onnx/defs/training/defs.cc

```diff
@@ -0,0 +1,171 @@
+                      ^
+                      |
+                      Z
+```
+
+, where W and Z are trainable tensors. Note that operators' attributes are
+omitted for the sake of simplicity. Let dY/dW (dY/dZ) be the gradient of
+Y with respect to W (Z). The user can compute gradient by inserting Gradient
+operator to form
+
+```
+W --> Conv --> H --> Gemm --> Y
+|      ^              ^
+|      |              |
+|      X              Z
+|      |              |
+|      |   .----------'
+|      |   |  (X/W/Z is the 1st/2nd/3rd input of Gradient as shown in "xs")
+|      v   v
+'---> Gradient(xs=["X", "W", "Z"], y=["Y"]) ---> dY/dX (1st output of Gradient)
+       |   |
+       |   '-----------------------------------> dY/dW (2nd output of Gradient)
+       |
+       '---------------------------------------> dY/dZ (3rd output of Gradient)
+```
+
+, where the content inside the braces of Gradient are attributes. The attribute
+"xs" specifies the necessary inputs to compute the variable specified by
+attribute "y". To compute gradient, the runtime should identify a sub-graph with
+inputs specified by "xs" and outputs containing "y"'s content and do a backward
+pass (or other auto-differentiation techniques whenever needed).
+
+Since X is not trainable tensor, the user can avoid the creation of dY/dX by
+assigning an empty string to the 1st output name of that Gradient. In other
+words, all Gradient's outputs are optional. Note that the concept of optional
+outputs can also be found in ONNX's RNN, GRU, and LSTM.
+
+Gradient operator can compute derivative against intermediate tensors. For
+example, the gradient of Y with respect to H can be done in
+
+```
+W --> Conv --> H --> Gemm --> Y
+       ^       |      ^
+       |       |      |
+       X       |      Z
+       .-------'      |
+       |   .----------'
+       |   | (H/Z is the 1st/2nd input of Gradient as shown in "xs")
+       v   v
+      Gradient(xs=["H", "Z"], y=["Y"])
+       |   |
+       |   '-----------------------------------> dY/dH (1st output of Gradient)
+       |
+       '---------------------------------------> dY/dZ (2nd output of Gradient)
+```
+
+It is possible to represent high-order differentiation using Gradient operator.
+An example for linear model is
+
+```
+W --> Gemm --> Y --> Loss --> O
+       ^              ^
+       |              |
+       X              L
+```
+
+To compute the 2nd order derivative of O with respect to W (denoted by
+d^2O/dW^2), one can do
+
+```
+W --> Gemm --> Y --> Loss --> O
+|      ^              ^
+|      |              |
+|      X              L
+|      |
+|      |
++------+--> Gradient(xs=["X", "W"], y=["Y"]) ---> dO/dX (1st output of Gradient)
+|      |      |
+|      |      '---> dO/dW (2nd output of Gradient)
+|      v
+'---> Gradient(xs=["X", "W"], y=["dO/dW"]) ---> d(dO/dW)dX (1st output of
+       |                                          Gradient)
+       |
+       |
+       '---> d^2O/dW^2 (2nd output of Gradient)
```

Comment:
Suggest using a valid identifier (as above).

Reply:
Humm, but `d(dO/dW)dX` would be changed to something like `d_dO_dW_dX` and become less readable.

Reply:
I see your point. But my concern is that some users will copy this blindly and use "dO/dW" as a tensor name. This can lead to potential problems downstream. Technically, the ONNX documentation says that tensor names should be valid C identifiers. If the ONNX checker were to enforce this, that would be somewhat helpful. But it does not seem to do that. (I remember Niklas was trying to add such a check ... but it doesn't seem to exist today … not sure why. Possibly because a number of models were violating this then.) We should probably sort this out one way or the other in the ONNX standard. Until then, I think it would be better to try to stick to the identifier restriction. 

We could use simply use names like D1 and D2 to denote the first and second derivative in the example (and in the text we could say D1 denotes dO/dW, for example).